### ISSUE TITLE
Labeled breaks / continues are transformed incorrectly inside lexical scope IIFE

### ISSUE THREAD

**Issue by Arnavion** [6to5/6to5#644]
That is, they aren't transformed at all.

Original:
```javascript
(function () {
  loop1:
  for (let j = 0; j < 10; j++) {
    for (let i = 0; i < 10; i++) {
      // Force an IIFE to capture i
      setTimeout(function () { console.log(i); }, 0);

      break loop1;
      continue loop1;
    }
  }
})();
```

Transformed:
```javascript
"use strict";

(function () {
  loop1: for (var j = 0; j < 10; j++) {
    for (var i = 0; i < 10; i++) {
      (function (i) {
        // Force an IIFE to capture i
        setTimeout(function () {
          console.log(i);
        }, 0);
        break loop1;
        continue loop1;
      })(i);
    }
  }
})();
```

The loop1 label is invalid inside the IIFE. It needs a similar transformation to return or unlabeled break and continue, eg:

```javascript
"use strict";

(function () {
  loop1: for (var j = 0; j < 10; j++) {
    for (var i = 0; i < 10; i++) {
      var _ret = (function (i) {
        // Force an IIFE to capture i
        setTimeout(function () {
          console.log(i);
        }, 0);
        return "break|loop1";
        return "continue|loop1";
      })(i);

      switch (_ret) {
        case "break|loop1":
          break loop1;
        case "continue|loop1":
          continue loop1;
      }
    }
  }
})();
```

**Comment by sebmck** [comment]
I was going to actually revamp the way those return values were handled and just use an array with numbers and corresponding values. I'll attempt it now and get back to you.

**Comment by Arnavion** [comment]
Sure, that'll work too.

It might hurt readability a bit because a reader wouldn't know what a value meant until they saw its case in the switch block. You could put block comments on the individual returns to indicate the original meaning (like `return /* break loop1*/ 0; `) to help with that.

**Comment by sebmck** [comment]
Alright, I didn't do the optimisations that I mentioned but I've revamped the way `BreakStatement`s and `ContinueStatement`s are handled so there shouldn't be any unnecessary transformations and it recurses into loops and remaps labels too with the method you suggested. I'll have to do some additional testing as I can think of a few edgecases that wont be caught.

**Comment by sebmck** [comment]
Nope, turns out my implementation catches it all. The code is a bit yuck though on really complicated nested labels, I can work on optimising it later down the road.

**Comment by Arnavion** [comment]
Yep, if the labeled break/continue is inside more than one nested IIFE from the label then you would need to keep propogating it until the appropriate level.

I can't test it for myself at the moment, but your test/fixtures/transformation/es6-block-scoping-exec/nested-labels.js test seems to be that very scenario (continue through two IIFEs), so :+1:

**Comment by sebmck** [comment]
Released as of 3.2.0, thanks!

**Comment by Arnavion** [comment]
@sebmck Just looked at the new transform, and break is transformed the same as continue.

```javascript
(function () {
  var stack = [];

  loop1:
  for (let j = 0; j < 10; j++) {
    for (let i = 0; i < 10; i++) {
      stack.push(() => [i, j]);
      break loop1;
    }
  }

  return stack;
})();

// [0, 0]
```

```javascript
"use strict";

(function () {
  var stack = [];

  loop1: for (var j = 0; j < 10; j++) {
    (function (j) {
      for (var i = 0; i < 10; i++) {
        var _ret2 = (function (i) {
          stack.push(function () {
            return [i, j];
          });
          return {
            v: "break|loop1"
          };
        })(i);

        if (typeof _ret2 === "object") return _ret2.v;
      }
    })(j);
  }

  return stack;
})();

// [0, 0], [0, 1], ... [0, 9] - identical output as continue
```

Not sure if the continue implementation is missing an explicit `continue loop1;` also - the returns make it work so maybe you did that on purpose. But break definitely needs it.

**Comment by sebmck** [comment]
@Arnavion Looks like the return value checks are getting lost somewhere. I'll need to wrap my head around the transformations as they're quite fragile.

**Comment by sebmck** [comment]
Tracked it down. The "map" that we're creating wasn't properly being propagated back up. Fixed in 3.2.1. Any additional tests and examples, even if they pass, is extremely appreciated.
